### PR TITLE
[llm benchmark] Add flash attention 2

### PR DESCRIPTION
**Pull request by w5688414**
####  TODO

+ [x] Attention is None is not reliable
+ [x] need to wait for flash attention2 PR merge

```
export CUDA_VISIBLE_DEVICES=4
python benchmark.py --model_name_or_path bigscience/bloomz-7b1-mt  \
    --num_train_epochs 1 --per_device_train_batch_size 2 \
    --lora \
    --use_flash_attention \
    --intokens \
    --intokens_length 1024 \
    --train_data_size 1000 \
    --evaluation_strategy no --save_strategy no \
    --fp16 --fp16_opt_level O2 \
    --logging_steps 50 --output_dir outputs


export CUDA_VISIBLE_DEVICES=4
python benchmark.py --model_name_or_path bigscience/bloomz-7b1-mt  \
    --num_train_epochs 1 --per_device_train_batch_size 2 \
    --lora \
    --use_flash_attention \
    --intokens \
    --intokens_length 1024 \
    --train_data_size 1000 \
    --evaluation_strategy no --save_strategy no \
    --bp16 --fp16_opt_level O2 \
    --logging_steps 50 --output_dir outputs
```

### CODE REVIEW

File: paddlenlp/transformers/bloom/configuration.py [hunk 1]
```diff
@@ -148,3 +149,4 @@ def __init__(
         self.slow_but_exact = slow_but_exact
         self.use_recompute = use_recompute
         self.use_pure_fp16 = use_pure_fp16
+        self.use_flash_attention = use_flash_attention
```
Comment (sijunhe) [inline paddlenlp/transformers/bloom/configuration.py:152]: 这个字段放到基类里面吧，默认为false

File: paddlenlp/transformers/chatglm/modeling.py [hunk 2]
```diff
@@ -264,66 +264,94 @@ def forward(
...
+            k_layer = paddle.transpose(k_layer, [1, 0, 2, 3])
+            v_layer = paddle.transpose(v_layer, [1, 0, 2, 3])
+            query_states, key_states, value_states = q_layer, k_layer, v_layer
+
+            attn_output = F.scaled_dot_product_attention(
+                query_states,
+                key_states,
+                value_states,
+                attn_mask=attention_mask,
+                is_causal=attention_mask is None,
```
Comment (sijunhe) [inline paddlenlp/transformers/chatglm/modeling.py]: chatglm不可能是causal mask, 直接is_causal=False

File: paddlenlp/transformers/llama/modeling.py [hunk 3]
```diff
@@ -195,13 +195,15 @@ def scaled_dot_product_attention(
         if alibi is not None:
             raise ValueError("Flash Attention does not support ALiBi yet")
```
Comment (sijunhe) [inline paddlenlp/transformers/llama/modeling.py]: 这里也需要支持一下，不能再直接报错了

File: paddlenlp/transformers/llama/modeling.py [hunk 4]
```diff
@@ -195,13 +195,15 @@ def scaled_dot_product_attention(
...
-        attn_output, attn_weights = flash_attention(
+        # TODO: attention_mask is None is not reliable
+        attn_output = F.scaled_dot_product_attention(
             query_states,
             key_states,
             value_states,
-            causal=is_causal and query_states.shape[1] != 1,
-            return_softmax=output_attentions,
+            attn_mask=attention_mask,
+            is_causal=attention_mask is None,
```
Comment (sijunhe) [inline paddlenlp/transformers/llama/modeling.py:202]: 这样子不行，会和上次一样，让预训练也走入带mask的FA. 我和阿辉商量了，这块我来提一个PR处理一下，你基于这个修改吧

File: paddlenlp/transformers/chatglm/modeling.py (hunk 2 above)
Comment (w5688414) [reply paddlenlp/transformers/chatglm/modeling.py]: 已修改

File: paddlenlp/transformers/llama/modeling.py (hunk 3 above)
Comment (w5688414) [reply paddlenlp/transformers/llama/modeling.py]: 已修改

File: paddlenlp/transformers/llama/modeling.py [hunk 5]
```diff
@@ -193,15 +193,16 @@ def scaled_dot_product_attention(
...
+            attention_mask = attention_mask.cast(alibi.dtype) + alibi
+        # TODO: attention_mask is None is not reliable
+        attn_output = F.scaled_dot_product_attention(
             query_states,
             key_states,
             value_states,
-            causal=is_causal and query_states.shape[1] != 1,
-            return_softmax=output_attentions,
+            attn_mask=attention_mask,
+            is_causal=config.is_causal,
```
Comment (sijunhe) [inline paddlenlp/transformers/llama/modeling.py]: 如果真的要is_causal这个，是不是应该用185行的is_causal传入进来？
@ZHUI 一起来看一下这个地方, 之前我考虑过预训练不传入attention_mask来控制FA的is_causal, 但是当前的写法是如果不传入attention_mask的话，https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/llama/modeling.py#L1223 这个地方会先生成2D mask再expand成4D, 到了scaled_dot_product这里就无法区分了

File: paddlenlp/transformers/llama/modeling.py [hunk 6]
```diff
@@ -193,15 +193,16 @@ def scaled_dot_product_attention(
         # Paddle Flash Attention input [ bz, seqlen, nhead, head_dim]
         # Torch Flash Attention input [ bz, nhead, seqlen, head_dim]
         if alibi is not None:
-            raise ValueError("Flash Attention does not support ALiBi yet")
-
-        attn_output, attn_weights = flash_attention(
+            attention_mask = attention_mask.cast(alibi.dtype) + alibi
+        # TODO: attention_mask is None is not reliable
```
Comment (sijunhe) [inline paddlenlp/transformers/llama/modeling.py]: 如果is_causal, attention_mask是否应该赋值为None

**Review by sijunhe — CHANGES_REQUESTED** [review]
https://github.com/PaddlePaddle/PaddleNLP/tree/develop/llm/llama#3-%E9%A2%84%E8%AE%AD%E7%BB%83
llama的预训练也需要验证一下

File: paddlenlp/transformers/llama/modeling.py (hunk 5 above)
Comment (ZHUI) [reply paddlenlp/transformers/llama/modeling.py]: > https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/llama/modeling.py#L1223 这个地方会先生成2D

可以这里如果是`use_flash_attention`的话，不制作attention_mask吗？

File: paddlenlp/transformers/llama/modeling.py [hunk 7]
```diff
@@ -1243,6 +1250,10 @@ def forward(
         attention_mask = self._prepare_decoder_attention_mask(
             attention_mask, (batch_size, seq_length), cache_length, inputs_embeds.dtype
         )  # [bs, 1, seq_len, seq_len]
+
+        is_casual = is_casual_mask(attention_mask, inputs_embeds.dtype)
+        if is_casual:
```
Comment (ZHUI) [inline paddlenlp/transformers/llama/modeling.py]:
```suggestion
        if is_casual and self.use_flash_attention:
```

Comment (sijunhe) [inline paddlenlp/transformers/llama/modeling.py]: 这个地方应该只有self.use_flash_attention为true的时候，才需要is_casual_mask。不开FA可以不跑这个

File: paddlenlp/transformers/llama/modeling.py [hunk 8]
```diff
@@ -259,6 +260,12 @@ def masked_fill(x, mask, value):
     return paddle.where(mask, y, x)
 
 
+def is_casual_mask(attention_mask, dtype):
+    mask = paddle.tril(paddle.ones(attention_mask.shape))
+    inf_mask = paddle.where(mask, 0.0, paddle.finfo(dtype).min)
+    return (inf_mask == attention_mask).all().item()
```
Comment (sijunhe) [inline paddlenlp/transformers/llama/modeling.py]: 1. 当前的实现方法264行的dtype需要是bool, 要不然动转静报错，见单测
2. 这里的实现可以更简单些，不用重新造一个inf_mask. 可不可以直接`paddle.tril(attention_mask) == attention_mask`这个逻辑呢？

File: paddlenlp/transformers/llama/modeling.py (hunk 7 above)
Comment (w5688414) [reply paddlenlp/transformers/llama/modeling.py]: 我添加到is_casual_mask的上面了

File: paddlenlp/transformers/llama/modeling.py (hunk 8 above)
Comment (w5688414) [reply paddlenlp/transformers/llama/modeling.py]: 已修改

**Review by sijunhe — APPROVED** [review]
lgtm
